### DISCUSSION TITLE
Build -> Plugins -> ContentPlayer (Build Failing on various NodeJS Versions including the default version installed in Jenkins Server)

### DISCUSSION THREAD

**Discussion by shounakacharya** [Sunbird-Ed/Community#87] [Installation] [answered]
System Configuration Details:
1.	KP, DP, DB, Yarn, Jenkins, Other-Services: Ubuntu 18.04 LTS, 4vCPU, 16GB RAM, 250 GB HDD
2.	Google Kubernetes engine: 1.21.12-gke.2200 with 4 worker nodes. Note: Currently Sunbird is not compatible with K8s APIs  >= 1.22 and we have checked this
3.	All systems are running on GCP
4. GithubTag: release-4.6.0_RC1

It takes a lot of time and then spits out the error. Grunt is installed and it is taking npm 10.16.3

@gandham-santhosh and  @keshavprasadms

Jenkins Logs:

Started by user [Admin Deployer](http://35.222.148.131:8080/user/admin)
Rebuilds build #[2](http://35.222.148.131:8080/job/Build/job/Plugins/job/ContentPlayer/2)
Checking out git https://github.com/project-sunbird/sunbird-content-player.git into /var/lib/jenkins/workspace/Build/Plugins/ContentPlayer@script to read Jenkinsfile
The recommended git tool is: NONE
No credentials specified
 > git rev-parse --resolve-git-dir /var/lib/jenkins/workspace/Build/Plugins/ContentPlayer@script/.git # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url https://github.com/project-sunbird/sunbird-content-player.git # timeout=10
Fetching upstream changes from https://github.com/project-sunbird/sunbird-content-player.git
 > git --version # timeout=10
 > git --version # 'git version 2.17.1'
 > git fetch --tags --progress -- https://github.com/project-sunbird/sunbird-content-player.git +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git rev-parse origin/release-4.6.0_RC1^{commit} # timeout=10
 > git rev-parse release-4.6.0_RC1^{commit} # timeout=10
Checking out Revision 053d85386940cc30b5ce07f2d953efe71ae88e8c (release-4.6.0_RC1)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 053d85386940cc30b5ce07f2d953efe71ae88e8c # timeout=10
Commit message: "Merge pull request #495 from gouravmore/release-4.4.0"
 > git rev-list --no-walk 053d85386940cc30b5ce07f2d953efe71ae88e8c # timeout=10
Running in Durability level: MAX_SURVIVABILITY
[Pipeline] Start of Pipeline
[Pipeline] node
Running on [Jenkins](http://35.222.148.131:8080/computer/(built-in)/) in /var/lib/jenkins/workspace/Build/Plugins/ContentPlayer
[Pipeline] {
[Pipeline] ansiColor
[Pipeline] {

[Pipeline] stage
[Pipeline] { (Checkout)
[Pipeline] cleanWs
[WS-CLEANUP] Deleting project workspace...
[WS-CLEANUP] Deferred wipeout is used...
[WS-CLEANUP] done
[Pipeline] checkout
The recommended git tool is: NONE
No credentials specified
Cloning the remote Git repository
Cloning repository https://github.com/project-sunbird/sunbird-content-player.git
 > git init /var/lib/jenkins/workspace/Build/Plugins/ContentPlayer # timeout=10
Fetching upstream changes from https://github.com/project-sunbird/sunbird-content-player.git
 > git --version # timeout=10
 > git --version # 'git version 2.17.1'
 > git fetch --tags --progress -- https://github.com/project-sunbird/sunbird-content-player.git +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git config remote.origin.url https://github.com/project-sunbird/sunbird-content-player.git # timeout=10
 > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
Avoid second fetch
 > git rev-parse origin/release-4.6.0_RC1^{commit} # timeout=10
 > git rev-parse release-4.6.0_RC1^{commit} # timeout=10
Checking out Revision 053d85386940cc30b5ce07f2d953efe71ae88e8c (release-4.6.0_RC1)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 053d85386940cc30b5ce07f2d953efe71ae88e8c # timeout=10
Commit message: "Merge pull request #495 from gouravmore/release-4.4.0"
[Pipeline] sh
+ git rev-parse --short HEAD
[Pipeline] sh
+ rev
+ git name-rev --name-only HEAD
+ cut -d / -f1
+ rev
[Pipeline] sh
+ echo release-4.6.0_RC1_053d85386_4
[Pipeline] echo
artifact_version: release-4.6.0_RC1_053d85386_4
[Pipeline] stage
[Pipeline] { (Build)
[Pipeline] sh
v10.16.3 is already installed.
Now using node v10.16.3 (npm v6.9.0)
npm WARN deprecated grunt-rename@0.1.4: Deprecated
npm WARN deprecated phantomjs-prebuilt@2.1.16: this package is now deprecated
npm WARN deprecated uglify-es@3.3.9: support for ECMAScript is superseded by `uglify-js` as of v3.13.0
npm WARN deprecated purifycss-webpack@0.7.0: Use purgecss-webpack-plugin instead
npm WARN deprecated fingerprintjs2@2.1.4: Package has been renamed to @fingerprintjs/fingerprintjs. Install @fingerprintjs/fingerprintjs to get updates.
npm WARN deprecated xmldom@0.1.31: Deprecated due to CVE-2021-21366 resolved in 0.5.0
npm WARN deprecated minimatch@0.2.14: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
npm WARN deprecated coffee-script@1.3.3: CoffeeScript on NPM has moved to "coffeescript" (no hyphen)
npm WARN deprecated node-uuid@1.4.8: Use uuid module instead
npm WARN deprecated iltorb@2.4.5: The zlib module provides APIs for brotli compression/decompression starting with Node.js v10.16.0, please use it over iltorb
npm WARN deprecated gulp-util@3.0.8: gulp-util is deprecated - replace it, following the guidelines at https://medium.com/gulpjs/gulp-util-ca3b1f9f9ac5
npm WARN deprecated chokidar@1.7.0: Chokidar 2 will break on node v14+. Upgrade to chokidar 3 with 15x less dependencies.
npm WARN deprecated core-js@2.6.12: core-js@<3.23.3 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up to 100x even if nothing is polyfilled. Some versions have web compatibility issues. Please, upgrade your dependencies to the actual version of core-js.
npm WARN deprecated log4js@0.6.38: 0.x is no longer supported. Please upgrade to 6.x or higher.
npm WARN deprecated istanbul@0.4.5: This module is no longer maintained, try this instead:
npm WARN deprecated   npm i nyc
npm WARN deprecated Visit https://istanbul.js.org/integrations for other alternatives.
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated nomnom@1.8.1: Package no longer supported. Contact support@npmjs.com for more info.
npm WARN deprecated connect@2.30.2: connect 2.x series is deprecated
npm WARN deprecated mkdirp@0.5.1: Legacy versions of mkdirp are no longer supported. Please update to mkdirp 1.x. (Note that the API surface has changed to use Promises in 1.x.)
npm WARN deprecated hoek@4.2.1: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated joi@13.7.0: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated graceful-fs@1.2.3: please upgrade to graceful-fs 4 for compatibility with current and future versions of Node.js
npm WARN deprecated request@2.47.0: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated tar@1.0.2: This version of tar is no longer supported, and will not receive security updates. Please upgrade asap.
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated svgo@1.3.2: This SVGO version is no longer supported. Upgrade to v2.x.x.
npm WARN deprecated fsevents@1.2.13: fsevents 1 will break on node v14+ and could be using insecure binaries. Upgrade to fsevents 2.
npm WARN deprecated stable@0.1.8: Modern JS already guarantees Array#sort() is a stable sort, so this library is deprecated. See the compatibility table on MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#browser_compatibility
npm WARN deprecated jasmine-matchers-loader@0.1.0: jasmine-matchers-loader has been deprecated, please migrate to https://github.com/JamieMason/add-matchers
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated babel-preset-es2015@6.24.1: 🙌  Thanks for using Babel: we recommend using babel-preset-env now: please read https://babeljs.io/env to update!
npm WARN deprecated chokidar@2.1.8: Chokidar 2 does not receive security updates since 2019. Upgrade to chokidar 3 with 15x fewer dependencies
npm WARN deprecated log4js@3.0.6: 3.x is no longer supported. Please upgrade to 6.x or higher.
npm WARN deprecated hoek@5.0.4: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated uuid@2.0.3: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated minimatch@0.3.0: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
npm WARN deprecated npmconf@2.1.3: this package has been reintegrated into npm and is now out of date with respect to npm
npm WARN deprecated topo@3.0.3: This module has moved and is now available at @hapi/topo. Please update your dependencies as this version is no longer maintained an may contain bugs and security issues.
npm WARN deprecated hawk@1.1.1: This module moved to @hapi/hawk. Please make sure to switch over as this distribution is no longer supported and may contain bugs and critical security issues.
npm WARN deprecated request@2.74.0: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated tar@2.2.2: This version of tar is no longer supported, and will not receive security updates. Please upgrade asap.
npm WARN deprecated uuid@3.0.1: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated minimatch@2.0.10: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
npm WARN deprecated natives@1.1.6: This module relies on Node.js's internals and will break at some point. Do not use it, and update to graceful-fs@4.x.
npm WARN deprecated json3@3.3.2: Please use the native JSON object instead of JSON 3
npm WARN deprecated browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
npm WARN deprecated svgo@0.7.2: This SVGO version is no longer supported. Upgrade to v2.x.x.
npm WARN deprecated source-map-resolve@0.5.3: See https://github.com/lydell/source-map-resolve#deprecated
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated querystring@0.2.0: The querystring API is considered Legacy. new code should use the URLSearchParams API instead.
npm WARN deprecated minimatch@1.0.0: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
npm WARN deprecated node-uuid@1.3.3: Use uuid module instead
npm WARN deprecated ini@1.1.0: Please update to ini >=1.3.6 to avoid a prototype pollution issue
npm WARN deprecated circular-json@0.5.9: CircularJSON is in maintenance only, flatted is its successor.
npm WARN deprecated date-format@1.2.0: 1.x is no longer supported. Please upgrade to 4.x or higher.
npm WARN deprecated hoek@6.1.3: This module has moved and is now available at @hapi/hoek. Please update your dependencies as this version is no longer maintained an may contain bugs and security issues.
npm WARN deprecated streamroller@0.7.0: 0.x is no longer supported. Please upgrade to 3.x or higher.
npm WARN deprecated hoek@0.9.1: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated sntp@0.2.4: This module moved to @hapi/sntp. Please make sure to switch over as this distribution is no longer supported and may contain bugs and critical security issues.
npm WARN deprecated boom@0.4.2: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated cryptiles@0.2.2: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated request@2.81.0: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated har-validator@2.0.6: this library is no longer supported
npm WARN deprecated hawk@3.1.3: This module moved to @hapi/hawk. Please make sure to switch over as this distribution is no longer supported and may contain bugs and critical security issues.
npm WARN deprecated flatten@1.0.3: flatten is deprecated in favor of utility frameworks such as lodash.
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated source-map-url@0.4.1: See https://github.com/lydell/source-map-url#deprecated
npm WARN deprecated har-validator@4.2.1: this library is no longer supported
npm WARN deprecated hoek@2.16.3: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated boom@2.10.1: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated cryptiles@2.0.5: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated sntp@1.0.9: This module moved to @hapi/sntp. Please make sure to switch over as this distribution is no longer supported and may contain bugs and critical security issues.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.2.7 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.13: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

npm ERR! Error while executing:
npm ERR! /usr/bin/git ls-remote -h -t git://github.com/bessdsv/bower-installer.git
npm ERR! 
npm ERR! undefined
npm ERR! exited with error code: 128

npm ERR! A complete log of this run can be found in:
npm ERR!     /var/lib/jenkins/.npm/_logs/2022-08-19T08_26_18_112Z-debug.log

> @project-sunbird/content-player@4.6.0 build-preview /var/lib/jenkins/workspace/Build/Plugins/ContentPlayer/player
> bash -c 'grunt uglify:renderermin && if [[ ${1} == "ekstep" ]]; then npm run package-coreplugins -- --env.channel=ekstep && npm run build-ekstep && grunt build-preview; else echo ${1} npm run package-coreplugins -- --env.channel=sunbird && npm run build-sunbird && grunt build-preview; fi' --  "ekstep"

grunt-cli: The grunt command line interface (v1.2.0)

Fatal error: Unable to find local grunt.

If you're seeing this message, grunt hasn't been installed locally to
your project. For more information about installing and configuring grunt,
please see the Getting Started guide:

http://gruntjs.com/getting-started
npm ERR! code ELIFECYCLE
npm ERR! errno 99
npm ERR! @project-sunbird/content-player@4.6.0 build-preview: `bash -c 'grunt uglify:renderermin && if [[ ${1} == "ekstep" ]]; then npm run package-coreplugins -- --env.channel=ekstep && npm run build-ekstep && grunt build-preview; else echo ${1} npm run package-coreplugins -- --env.channel=sunbird && npm run build-sunbird && grunt build-preview; fi' --  "ekstep"`
npm ERR! Exit status 99
npm ERR! 
npm ERR! Failed at the @project-sunbird/content-player@4.6.0 build-preview script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /var/lib/jenkins/.npm/_logs/2022-08-19T08_26_18_628Z-debug.log
grunt-cli: The grunt command line interface (v1.2.0)

Fatal error: Unable to find local grunt.

If you're seeing this message, grunt hasn't been installed locally to
your project. For more information about installing and configuring grunt,
please see the Getting Started guide:

http://gruntjs.com/getting-started
grunt-cli: The grunt command line interface (v1.2.0)

Fatal error: Unable to find local grunt.

If you're seeing this message, grunt hasn't been installed locally to
your project. For more information about installing and configuring grunt,
please see the Getting Started guide:

http://gruntjs.com/getting-started
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // stage
[Pipeline] }

[Pipeline] // ansiColor
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
ERROR: script returned exit code 99
Finished: FAILURE

**Comment by beepdot** [comment, marked as answer]
@vinukumar-vs Please can you suggest a solution for the error?

**Reply by HarishGangula** [reply]
@shounakacharya  This is fixed in https://github.com/project-sunbird/sunbird-content-player/tree/release-4.6.0. Please use this branch.
cc: @keshavprasadms

**Reply by vinukumar-vs** [reply]
@shounakacharya Refer this PR for Build issue fixes
https://github.com/project-sunbird/sunbird-content-player/pull/499/files

**Reply by shounakacharya** [reply]
@HarishGangula @vinukumar-vs @keshavprasadms : Thanks. This worked. Marking this as answered!
